test/dialects/spark3_grammar_test.py

```diff
@@ -0,0 +1,97 @@
+    [
+        (t[1], b, t[0])
+        for t in [
+            ("as t ( col1 , col2 )", "AliasExpressionSegment"),
+            ("as t \t ( col1 , col2 )", "AliasExpressionSegment"),
+            ("from t", "FromClauseSegment"),
+            ("from t as u", "FromClauseSegment"),
+            ("from values ( 1 , 2 )", "FromClauseSegment"),
+            ("from values 1", "FromClauseSegment"),
+            ("from values 1 , 2 ", "FromClauseSegment"),
+            ("1 , 2 , 3", "DelimitedValues"),
+            ("values 1 , 2 , 3", "ValuesClauseSegment"),
+            ("values ( 1 , 2 , 3 )", "ValuesClauseSegment"),
+            ("values ( 1 , 2 ) , ( 3 , 4 )", "ValuesClauseSegment"),
+            ("values ( 1 , 2 )", "TableExpressionSegment"),
+            ("values ( 1 , 2 ) , ( 3 , 4 )", "TableExpressionSegment"),
+            ("values ( 1 , 2 )", "FromExpressionElementSegment"),
+            ("values ( 1 , 2 ) , ( 3 , 4 )", "FromExpressionElementSegment"),
+            ("values ( 1 , 2 ) , ( 3 , 4 )", "FromExpressionSegment"),
+            ("from values ( 1 , 2 )", "FromClauseSegment"),
+            ("from values 1 , 2", "FromClauseSegment"),
+            ("from values ( 1 , 2 ) , ( 3 , 4 )", "FromClauseSegment"),
+            ("select * from t", "SelectStatementSegment"),
+            ("select * from values ( 1 , 2 ) , ( 3 , 4 ) ;", "SelectStatementSegment"),
+        ]
+        for b in ["match", "parse"]
+    ],
+)
+def test__dialect__spark3__grammars(
+    segment_class: str,
+    match_or_parse: str,
+    fragment: str,
+    generate_test_segments: Callable[[List[str]], List[RawSegment]],
+    fresh_spark3_dialect: Dialect,
+    caplog: Any,
+):
+    """Test that some grammars match as expected.
+
+    When the fixtured tests fail to parse some particular query, it can be difficult to tell where things went wrong.
+    For instance, "values (1, 2)" is a `ValuesClauseSegment` but also a `TableExpressionSegment`. If this fails to be
+    recognized, is it because `ValuesClauseSegment` is wrong, or because a `ValuesClauseSegment` is not recognized as
+    a `TableExpressionSegment`?
+
+    This test can help us diagnose that, since we pick both the sequence of segments _and_ the grammar we think it
+    should match.
+    """
+    lexed_seg_list = lex_segments(fragment)
+    segment = fresh_spark3_dialect._library[segment_class]
+    grammar = (
+        segment.parse_grammar
+        if (match_or_parse == "parse" and segment.parse_grammar)
+        else segment.match_grammar
+    )
+
+    with RootParseContext(dialect=fresh_spark3_dialect) as ctx:
+        ctx.logger.setLevel(logging.DEBUG)
+        with caplog.at_level(logging.DEBUG):
+            p = grammar.match(lexed_seg_list, parse_context=ctx)
+
+    if p.unmatched_segments:
+        logger = logging.getLogger()
+        logger.addHandler(logging.StreamHandler())
+        for _, _, message in caplog.record_tuples:
+            logger.warning(message)
```

Comment:
for my debugging purposes. 

i wanted to ask, is there a good place that explains how to read the detailed parsing output?

just by staring at it i could see it was only picking up part of `from values 1,2` as the `FromExpressionSegment`, but beyond that i was having trouble reading it:

```
[PD:0  MD:1 ]	.None                                             	Ref.match OUT       	+   	[match=<MatchResult 1/9: 'from'>, seg="'from values 1 , 2'"]
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match IN	    	[ls=7, seg='values 1 , 2']
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match SI	    	[mq=[], sb=['VALUES', '', '1', '', ',', '', '2']]
[PD:0  MD:2 ]	..None                                            	Ref.match OUT       	+   	[match=<MatchResult 1/3: ','>, seg="', 2'"]
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match SC	    	[bsm=(4, 1, <Ref: CommaSegment>)]
[PD:0  MD:4 ]	....FromExpressionSegment                         	OneOf.match PRN     	    	[ns=1, ps=1, ms=0, pruned=[<Ref: MLTableExpressionSegment>], opts=[<Ref: FromExpressionElementSegment>]]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Ref.match OUT       	    	[match=<MatchResult 0/3: ''>, seg="'values 1'"]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Optio.match PRN     	    	[ns=1, ps=1, ms=0, pruned=[<Bracketed: [<Ref: TableExpressionSegment>]>], opts=[<Ref: TableExpressionSegment>]]
[PD:0  MD:9 ]	.........TableExpressionSegment                   	OneOf.match PRN     	    	[ns=3, ps=1, ms=1, pruned=[<Bracketed: [<Ref: SelectableGrammar>]>], opts=[<Ref: ValuesClauseSegment>, <Ref: BareFunctionSegment>, <Ref: FunctionSegment>, <Ref: TableReferenceSegment>]]
[PD:0  MD:12]	............ValuesClauseSegment                   	Ref.match OUT       	+   	[match=<MatchResult 1/3: 'values'>, seg="'values 1'"]
[PD:0  MD:11]	...........ValuesClauseSegment                    	Start.match OUT     	++  	[match=<MatchResult 3/3: 'values 1'>, seg="'values 1'"]
[PD:0  MD:10]	..........ValuesClauseSegment                     	Value.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:10]	..........TableExpressionSegment                  	Ref.match OUT       	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:9 ]	.........TableExpressionSegment                   	OneOf.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:8 ]	........TableExpressionSegment                    	Table.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:8 ]	........FromExpressionElementSegment              	Ref.match OUT       	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Optio.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:6 ]	......FromExpressionElementSegment                	Seque.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:5 ]	.....FromExpressionElementSegment                 	FromE.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:5 ]	.....FromExpressionSegment                        	Ref.match OUT       	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:4 ]	....FromExpressionSegment                         	OneOf.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:3 ]	...FromExpressionSegment                          	Condi.match OUT     	++  	[match=<MatchResult 1/1: ''>, seg="''"]
[PD:0  MD:3 ]	...FromExpressionSegment                          	Seque.match OUT     	++  	[match=<MatchResult 3/3: 'values 1'>, seg="'values 1'"]
[PD:0  MD:2 ]	..FromExpressionSegment                           	FromE.match OUT     	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:2 ]	..None                                            	Ref.match OUT       	++  	[match=<MatchResult 1/1: 'values 1'>, seg="'values 1'"]
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match IN	    	[ls=2, seg=' 2']
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match SI	    	[mq=[], sb=['', '2']]
[PD:0  MD:2 ]	..None                                            	Delim._look_ahead_match SC	    	[bsm=None]
[PD:0  MD:4 ]	....FromExpressionSegment                         	OneOf.match PRN     	    	[ns=1, ps=1, ms=0, pruned=[<Ref: MLTableExpressionSegment>], opts=[<Ref: FromExpressionElementSegment>]]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Optio.match PRN     	    	[ns=1, ps=1, ms=0, pruned=[<Bracketed: [<Ref: TableExpressionSegment>]>], opts=[<Ref: TableExpressionSegment>]]
[PD:0  MD:9 ]	.........TableExpressionSegment                   	OneOf.match PRN     	    	[ns=3, ps=2, ms=0, pruned=[<Ref: ValuesClauseSegment>, <Bracketed: [<Ref: SelectableGrammar>]>], opts=[<Ref: BareFunctionSegment>, <Ref: FunctionSegment>, <Ref: TableReferenceSegment>]]
[PD:0  MD:10]	..........TableExpressionSegment                  	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:11]	...........FunctionSegment                        	OneOf.match PRN     	    	[ns=1, ps=1, ms=0, pruned=[<Sequence: [<Sequence: [<Ref: DatePartFunctionNameSe...]>], opts=[<Sequence: [<Sequence: [<AnyNumberOf: [<Ref: Functio..., <Ref: PostFunctionGrammar [opt]>]>]]
[PD:0  MD:17]	.................DatePartFunctionNameSegment      	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:16]	................DatePartFunctionNameSegment       	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:15]	...............DatePartFunctionNameSegment        	DateP.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:15]	...............FunctionSegment                    	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:14]	..............FunctionSegment                     	AnyNu.match PRN     	    	[ns=1, ps=0, ms=0, pruned=[], opts=[<Ref: FunctionNameSegment>]]
[PD:0  MD:17]	.................FunctionNameSegment              	AnyNu.match PRN     	    	[ns=1, ps=0, ms=0, pruned=[], opts=[<Sequence: [<Ref: SingleIdentifierGrammar>, <Ref: DotSegment>]>]]
[PD:0  MD:19]	...................SingleIdentifierGrammar        	OneOf.match PRN     	    	[ns=2, ps=0, ms=0, pruned=[], opts=[<Ref: NakedIdentifierSegment>, <Ref: QuotedIdentifierSegment>]]
[PD:0  MD:20]	....................SingleIdentifierGrammar       	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:20]	....................SingleIdentifierGrammar       	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:19]	...................SingleIdentifierGrammar        	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:19]	...................FunctionNameSegment            	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:18]	..................FunctionNameSegment             	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:17]	.................FunctionNameSegment              	AnyNu.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:17]	.................FunctionNameSegment              	OneOf.match PRN     	    	[ns=2, ps=0, ms=0, pruned=[], opts=[<Ref: FunctionNameIdentifierSegment>, <Ref: QuotedIdentifierSegment>]]
[PD:0  MD:18]	..................FunctionNameSegment             	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:18]	..................FunctionNameSegment             	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:17]	.................FunctionNameSegment              	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:16]	................FunctionNameSegment               	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:15]	...............FunctionNameSegment                	Funct.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:15]	...............FunctionSegment                    	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:14]	..............FunctionSegment                     	AnyNu.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:13]	.............FunctionSegment                      	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............FunctionSegment                       	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:11]	...........FunctionSegment                        	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:10]	..........FunctionSegment                         	Funct.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:10]	..........TableExpressionSegment                  	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............TableReferenceSegment                 	Delim._look_ahead_match IN	    	[ls=1, seg='2']
[PD:0  MD:12]	............TableReferenceSegment                 	Delim._look_ahead_match SI	    	[mq=[], sb=['2']]
[PD:0  MD:12]	............TableReferenceSegment                 	OneOf.match PRN     	    	[ns=1, ps=10, ms=0, pruned=[<Ref: OnKeywordSegment>, <Ref: AsKeywordSegment>, <Ref: UsingKeywordSegment>, <Ref: CommaSegment>, <Ref: CastOperatorSegment>, <Ref: StartSquareBracketSegment>, <Ref: StartBracketSegment>, <Ref: ColonSegment>, <Ref: DelimiterSegment>, <class 'sqlfluff.core.parser.segments.base.BracketedSegment'>], opts=[<Ref: BinaryOperatorGrammar>]]
[PD:0  MD:13]	.............BinaryOperatorGrammar                	OneOf.match PRN     	    	[ns=1, ps=3, ms=0, pruned=[<Ref: ArithmeticBinaryOperatorGrammar>, <Ref: StringBinaryOperatorGrammar>, <Ref: BooleanBinaryOperatorGrammar>], opts=[<Ref: ComparisonOperatorGrammar>]]
[PD:0  MD:14]	..............ComparisonOperatorGrammar           	OneOf.match PRN     	    	[ns=1, ps=9, ms=0, pruned=[<Ref: EqualsSegment>, <Ref: EqualsSegment_a>, <Ref: EqualsSegment_b>, <Ref: GreaterThanSegment>, <Ref: LessThanSegment>, <Ref: GreaterThanOrEqualToSegment>, <Ref: LessThanOrEqualToSegment>, <Ref: NotEqualToSegment_a>, <Ref: NotEqualToSegment_b>], opts=[<Ref: LikeOperatorSegment>]]
[PD:0  MD:15]	...............ComparisonOperatorGrammar          	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:14]	..............ComparisonOperatorGrammar           	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:14]	..............BinaryOperatorGrammar               	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:13]	.............BinaryOperatorGrammar                	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:13]	.............TableReferenceSegment                	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............TableReferenceSegment                 	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............TableReferenceSegment                 	NonCo.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............SingleIdentifierGrammar               	OneOf.match PRN     	    	[ns=2, ps=0, ms=0, pruned=[], opts=[<Ref: NakedIdentifierSegment>, <Ref: QuotedIdentifierSegment>]]
[PD:0  MD:13]	.............SingleIdentifierGrammar              	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:13]	.............SingleIdentifierGrammar              	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............SingleIdentifierGrammar               	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:12]	............TableReferenceSegment                 	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:11]	...........TableReferenceSegment                  	Delim.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:10]	..........TableReferenceSegment                   	Table.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:10]	..........TableExpressionSegment                  	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:9 ]	.........TableExpressionSegment                   	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:8 ]	........TableExpressionSegment                    	Table.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:8 ]	........FromExpressionElementSegment              	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:7 ]	.......FromExpressionElementSegment               	Optio.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:6 ]	......FromExpressionElementSegment                	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:5 ]	.....FromExpressionElementSegment                 	FromE.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:5 ]	.....FromExpressionSegment                        	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:4 ]	....FromExpressionSegment                         	OneOf.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:3 ]	...FromExpressionSegment                          	Seque.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:2 ]	..FromExpressionSegment                           	FromE.match OUT     	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:2 ]	..None                                            	Ref.match OUT       	    	[match=<MatchResult 0/1: ''>, seg="'2'"]
[PD:0  MD:1 ]	.None                                             	Delim.match OUT     	    	[match=<MatchResult 0/2: ''>, seg="'values 1 , 2'"]
[PD:0  MD:0 ]	None                                              	Seque.match OUT     	    	[match=<MatchResult 0/9: ''>, seg="'from values 1 , 2'"]
```

Reply:
Nothing I'm aware of. @alanmcruickshank wrote the parser;  maybe he has something to add?